src/binding/c/request_api.txt

```diff
@@ -316,8 +326,12 @@ MPI_Testsome:
     for (int i = 0; i < *outcount; i++) {
         int idx = array_of_indices[i];
         if (request_ptrs[idx] && !MPIR_Request_is_persistent(request_ptrs[idx])) {
-            MPIR_Request_free(request_ptrs[idx]);
+            int rc = MPIR_Request_free_return(request_ptrs[idx]);
             array_of_requests[idx] = MPI_REQUEST_NULL;
+            if (rc && array_of_statuses != MPI_STATUSES_IGNORE) {
+                array_of_statuses[idx].MPI_ERROR = rc;
```

Comment:
are the other status objects guaranteed to have `MPI_SUCCESS` in them at this point?

Reply:
Ahh, a typo. `array_of_statuses[idx]` should be `array_of_statuses[i]`. 

> are the other status objects guaranteed to have MPI_SUCCESS in them at this point?

I am not sure I understand the question. Why would it matter?

Reply:
The user is supposed to look in the status array to find out which operation has an error. So presumably all operations that succeeded need to have `MPI_SUCCESS` in the status object or else the user may think that they failed? That way this actual error stands out as the one needing handling.

Reply:
Quote from 6.12 -
> Upon returning from a completion call in which a nonblocking collective operation
completes, the values of the MPI_SOURCE and MPI_TAG elds in the associated status object,
if any, are undened. The value of MPI_ERROR may be dened, if appropriate, according
to the specication in Section 3.2.5. It is valid to mix dierent request types (i.e., any
combination of collective requests, I/O requests, generalized requests, or point-to-point
requests) in functions that enable multiple completions (e.g., MPI_WAITALL). It is erroneous
to call MPI_REQUEST_FREE or MPI_CANCEL for a request associated with a nonblocking
collective operation. Nonblocking collective requests created using the APIs described in
this section are not persistent. However, persistent collective requests can be created using
persistent collective operations described in Sections 6.13 and 8.8.
> 
> Rationale.  Freeing an active nonblocking collective request could cause similar
problems as discussed for point-to-point requests (see Section 3.7.3). Cancelling a
request is not supported because the semantics of this operation are not well-dened.
(End of rationale.)

Reply:
quote from 3.2.5 -
> In general, message-passing calls do not modify the value of the error code eld of
status variables. This eld may be updated only by the functions in Section 3.7.5 which
return multiple statuses. The eld is updated if and only if such function returns with an
error code of MPI_ERR_IN_STATUS.
>
> Rationale. The error eld in status is not needed for calls that return only one status,
such as MPI_WAIT, since that would only duplicate the information returned by the
function itself. The current design avoids the additional overhead of setting it, in such
cases. The eld is needed for calls that return multiple statuses, since each request
may have had a dierent failure. (End of rationale.)

Reply:
I think the wording such as "in general" and "may be updated" does not exclude that we do set the `MPI_ERROR` field. @raffenet What do you think?

Reply:
Yes, I agree. I think we can always set it.

Reply:
I read the text again because the commit modifying a bunch of tests gave me pause. I no longer think we can just blindly set the `MPI_ERROR` field in status. This sentence seems clear cut:
> The field is updated if and only if such function returns with an error code of MPI_ERR_IN_STATUS.

Reply:
Okay, I'll reverse the changes. Still IMO, this is a nonsense clause.